File: Effuse.Server/Domain/Role.cs
namespace Effuse.Server.Domain;

public class Role
(
  Guid id,
  string name,
  DateTime created_on,
  List<Permission> permissions
)
{
  public Guid Id => id;

  public string Name => name;

  public DateTime CreatedOn => created_on;

  public List<Permission> Permissions => permissions;

  public bool HasPermission(PermissionRequest request)
  {
    return permissions.Any(p => p.Allows(request));
  }

  public Role WithPermission(Permission permission)
  {
    return new
    (
      id: id,
      name: name,
      created_on: created_on,
      permissions: [.. permissions, permission]
    );
  }
}

File: Effuse.UserInterface/domain/server-management/Role.ts
import { RolePermission } from "./RolePermission";

export type RoleProps = {
  id: string;
  name: string;
  permissions: Array<RolePermission>;
};

export class Role {
  readonly #props: RoleProps;

  constructor(props: RoleProps) {
    this.#props = props;
  }

  get Id() {
    return this.#props.id;
  }

  get Name() {
    return this.#props.name;
  }

  get Permissions() {
    return this.#props.permissions;
  }

  WithPermission(permission: RolePermission) {
    return new Role({
      ...this.#props,
      permissions: [...this.#props.permissions, permission],
    });
  }
}
